### PR TITLE
feat: support indivisible shards for TP model loading and TPlizing.

### LINKED ISSUE #37051: Incorrect calculation of strides leading to loss of param data upon tensor parallel use while sliced model loading
### System Info

- `transformers` version: 4.50.2
- Python version: 3.12.0
- Huggingface_hub version: 0.29.3
- Safetensors version: 0.5.3
- Accelerate version: 1.5.2
- Accelerate config:    not found
- DeepSpeed version: not installed
- PyTorch version (GPU?): 2.5.0+cu124 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: using TP
- Using GPU in script?: yes
- GPU type: NVIDIA A100-SXM4-80GB

### Who can help?

trainer: @muellerzr @SunMarc
TP: @ArthurZucker

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

You can load any model (For instance `ibm-granite/granite-3.1-2b-base`) with `tp_plan="auto"`. Make sure the vocab size is not divisible by world size (in this case 5) just for one sharding example, in this case we want to let TP shard lm_head with `colwise_rep`. You can then inspect size of the weight param of lm_head after application and it would be a multiple of world size however loses param data (in this case its downsized from 49152 to 49150). 

### Expected behavior

Loss of data / param parity should hold after application of TP. This is essentially happening due to `get_tensor_shard()` function https://github.com/huggingface/transformers/blob/348f3285c5114159d2ff4933b4b8ae36866d01a7/src/transformers/integrations/tensor_parallel.py#L120 in transformers. Specifically because of flooring of values for instance here - https://github.com/huggingface/transformers/blob/348f3285c5114159d2ff4933b4b8ae36866d01a7/src/transformers/integrations/tensor_parallel.py#L126. Essentially, PyTorch allows for unevening sharding with its placement APIs like Shard.

I am happy to help raise a PR if we agree on an approach.

### PR DESCRIPTION
**Pull request by kmehant**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/37051

Approach is to support uneven sharding and seek segments of data that mimics torch.chunk since torch.chunk is the style of sharding adopted by torch `Shard` placements API for both even and uneven sharding. Finally we pass stride and shape to `from_local` to allow for uneven sharding.

```python
from transformers import AutoModelForCausalLM
import torch

m2 = AutoModelForCausalLM.from_pretrained("TinyLlama/TinyLlama-1.1B-Chat-v1.0", tp_plan=None)
m = AutoModelForCausalLM.from_pretrained("TinyLlama/TinyLlama-1.1B-Chat-v1.0", tp_plan="auto")

print(m.model.layers[0].self_attn.q_proj.weight)
print(m.model.layers[0].self_attn.q_proj.weight.shape)
ft = m.model.layers[0].self_attn.q_proj.weight.full_tensor().to("cpu")
assert torch.equal(ft, m2.model.layers[0].self_attn.q_proj.weight.to("cpu"))
# assert should pass
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. https://github.com/huggingface/transformers/issues/37051
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

 @ArthurZucker @SunMarc @muellerzr

### CODE REVIEW

File: src/transformers/integrations/tensor_parallel.py [hunk 1]
```diff
@@ -120,13 +121,25 @@ def get_packed_weights(param, empty_param, device_mesh, rank, dim):
...
+        else:
+            param = torch.tensor([])
     elif dim == 2 or dim == -1:
         size_ = empty_param.shape[-1]
-        param = param[..., rank * (size_ // device_mesh.size()) : (rank + 1) * (size_ // device_mesh.size())]
+        split_size = math.ceil(size_ / device_mesh.size())
+        if (rank * split_size) < size_:
+            param = param[..., rank * split_size : min((rank + 1) * split_size, size_)]
+        else:
+            param = torch.tensor([])
```
Comment (ArthurZucker) [inline src/transformers/integrations/tensor_parallel.py:142]: not sure what happens in that case? IMO we should error out here no?

**Review by ArthurZucker — COMMENTED** [review]
thanks for opening a PR!

File: src/transformers/integrations/tensor_parallel.py (hunk 1 above)
Comment (kmehant) [reply src/transformers/integrations/tensor_parallel.py]: torch supports this case, here is the small reproducer. @ArthurZucker ⬇️ 

```python
import torch
from torch.distributed.tensor import init_device_mesh, DTensor, Shard
local_tensor = torch.tensor([1,2,3,4])

# Shard follows chunk semantics 
# For 6 ranks, below chunking would return only 4 tensors and other 2 tensors should be empty tensors of dtype int64
chunks = local_tensor.chunk(6)

# 6 ranked device mesh
device_mesh = init_device_mesh("cuda", (6,))

# we pass the chunk to each rank and empty tensor of int64 dtype for the last 2 ranks (4 and 5)
d = DTensor.from_local(chunks[torch.distributed.get_rank()] if torch.distributed.get_rank() < len(chunks) else torch.tensor([], dtype=torch.int64), device_mesh=device_mesh, placements=(Shard(0),), shape=local_tensor.size(), stride=local_tensor.stride())

# this prints something like below at each rank (local_tensor would change based on rank)
# DTensor(local_tensor=tensor([3], device='cuda:2'), device_mesh=DeviceMesh('cuda', [0, 1, 2, 3, 4, 5]), placements=(Shard(dim=0),))
# and something like below for remaining two ranks
# DTensor(local_tensor=tensor([], device='cuda:5', dtype=torch.int64), device_mesh=DeviceMesh('cuda', [0, 1, 2, 3, 4, 5]), placements=(Shard(dim=0),))
print(d)

# torch.Size([4])
print(d.size())

# We can materialize tensor back in each and every rank
# tensor([1, 2, 3, 4], device='cuda:0')
# tensor([1, 2, 3, 4], device='cuda:1')
# tensor([1, 2, 3, 4], device='cuda:2')
# tensor([1, 2, 3, 4], device='cuda:5')
# tensor([1, 2, 3, 4], device='cuda:3')
# tensor([1, 2, 3, 4], device='cuda:4')
# int64 was observed from how torch would chunk the tensors when using distribute_tensor() API, however, not using int64 dtype results in garbage values when materialized on the remaining 2 ranks.
print(d.full_tensor())
```

Comment (ArthurZucker) [reply src/transformers/integrations/tensor_parallel.py]: ah so nothing on that shardd?

Comment (kmehant) [reply src/transformers/integrations/tensor_parallel.py]: Yes, Arthur!

File: src/transformers/integrations/tensor_parallel.py [hunk 2]
```diff
@@ -141,13 +142,25 @@ def get_packed_weights(param, empty_param, device_mesh, rank, dim):
 def get_tensor_shard(param, empty_param, device_mesh, rank, dim):
     if dim == 0:
```
[the remark this answers is not in the document]
Comment (kmehant) [reply src/transformers/integrations/tensor_parallel.py:143]: Do you want me to document it as a code comment or a doc string to this function?

[the remark this answers is not in the document]
Comment (kmehant) [reply src/transformers/integrations/tensor_parallel.py:193]: Added the documentation.

**Comment by kmehant** [comment]
@ArthurZucker previous tests pass as well. Can you check review the updated implementation?

Screenshot 2025-04-11 at 2 16 02 PM

**Review by ArthurZucker — COMMENTED** [review]
Changes look good to me! 
Only thing left is.... adding a small test to show that a 1b llama can run with tp 3 for example! 🤗

**Comment by kmehant** [comment]
Screenshot 2025-04-11 at 9 20 07 PM

@ArthurZucker have added the test case. 


(not related to this PR) Could not do TP 3 on llama 1B since q_proj (https://github.com/huggingface/transformers/blob/bf46e44878bd86aebcfa1eceb4a93a6e5b20e863/src/transformers/models/llama/modeling_llama.py#L230) undergoes colwise instead of colwise_rep due to which the sharded output of q_proj wont be readily consumable here - https://github.com/huggingface/transformers/blob/bf46e44878bd86aebcfa1eceb4a93a6e5b20e863/src/transformers/models/llama/modeling_llama.py#L255

Its more of a tradeoff if we wish to do colwise_rep and get it working for any tp_size compromising on performance vs only one multiples of 2 (precisely head_dim to support view)

**Comment by kmehant** [comment]
Rebased the PR.

**Comment by kmehant** [comment]
@SunMarc @ArthurZucker Anything on this PR? :)

**Comment by kmehant** [comment]
@ArthurZucker @SunMarc Requesting review and merge, all conflicts resolved!

File: src/transformers/integrations/tensor_parallel.py [hunk 3]
```diff
@@ -139,15 +140,81 @@ def get_packed_weights(param, empty_param, device_mesh, rank, dim):
...
+    elif dim == 2:
+        if (rank * _split_size) < _dim_size:
+            param = param[:, :, rank * _split_size : min((rank + 1) * _split_size, _dim_size), ...]
+        else:
+            param = torch.empty((*empty_param.shape[:2], 0, *empty_param.shape[3:]), dtype=torch.int64)
+    elif dim == -1:
+        if (rank * _split_size) < _dim_size:
+            param = param[..., rank * _split_size : min((rank + 1) * _split_size, _dim_size)]
+        else:
+            param = torch.empty((*empty_param.shape[:-1], 0), dtype=torch.int64)
```
Comment (ArthurZucker) [inline src/transformers/integrations/tensor_parallel.py:217]: thanks and sorry was out for 2 weeks!
I think this can be refactored to me more readable no?! this is the only thing missing for me!

**Review by ArthurZucker — COMMENTED** [review]
thanks for updating

File: src/transformers/integrations/tensor_parallel.py (hunk 3 above)
Comment (kmehant) [reply src/transformers/integrations/tensor_parallel.py:217]: @ArthurZucker 
No worries, glad you took a break! Thanks for returning to this PR.

> I think this can be refactored to me more readable no?! this is the only thing missing for me!

Do you have any specific suggestions? Do you want me to use better variable names? and more assignments like keeping `rank * _split_size` in a variable and reuse something of that sort? 

Let me know, thank you.

Comment (kmehant) [reply src/transformers/integrations/tensor_parallel.py:217]: @ArthurZucker awaiting for you comments!

Comment (kmehant) [reply src/transformers/integrations/tensor_parallel.py:217]: sure @ArthurZucker please let me know, thanks

Comment (ArthurZucker) [reply src/transformers/integrations/tensor_parallel.py:217]: Merged!

Comment (kmehant) [reply src/transformers/integrations/tensor_parallel.py]: Thanks working on it!

**Comment by kmehant** [comment]
@ArthurZucker @SunMarc The PR is ready! Thank you.

**Comment by SunMarc** [comment]
cc @ArthurZucker

**Comment by kmehant** [comment]
@ArthurZucker requesting review on this PR.

**Comment by ArthurZucker** [comment]
Sorry all I was off for 2 weeks time to catchup

**Comment by kmehant** [comment]
@ArthurZucker thanks for circling back to this, I have rebased the PR and fixed the quality check that was failing.

**Comment by kmehant** [comment]
@ArthurZucker any update on this? We are holding to a fork, merging this PR, would help us (other TP users) to pivot on the HF transformers main repo. Thank you.

**Review by ArthurZucker — APPROVED** [review]
Sorry was waiting for the release to be passed, thanks a lot for your patience
